parlai/core/torch_agent.py

```diff
@@ -15,41 +15,54 @@
 from collections import deque, namedtuple
 import pickle
 import random
-import copy
 import math
 from operator import attrgetter
 
-Batch = namedtuple('Batch', [
-    # bsz x seqlen tensor containing the parsed text data
-    'text_vec',
-    # list of length bsz containing the lengths of the text in same order as
-    # text_vec; necessary for pack_padded_sequence
-    'text_lengths',
-    # bsz x seqlen tensor containing the parsed label (one per batch row)
-    'label_vec',
-    # list of length bsz containing the lengths of the labels in same order as
-    # label_vec
-    'label_lens',
-    # list of length bsz containing the selected label for each batch row (some
-    # datasets have multiple labels per input example)
-    'labels',
-    # list of length bsz containing the original indices of each example in the
-    # batch. we use these to map predictions back to their proper row, since
-    # e.g. we may sort examples by their length or some examples may be
-    # invalid.
-    'valid_indices',
-    # list of lists of tensors. outer list has size bsz, inner lists vary in
-    # size based on the number of candidates for each row in the batch.
-    'candidates',
-])
-
-Output = namedtuple('Output', [
-    # list of strings of length bsz containing the predictions of the model
-    'predictions',
-    # list of lists of length bsz containing the predictions of the model.
-    # each sub-list is an ordered ranking of strings of variable length.
-    'candidate_predictions',
-])
+"""
+Batch is a namedtuple containing data being sent to an agent.
+
+This is the input type of the train_step and eval_step functions.
+Agents can override the batchify function to return an extended namedtuple
+with additional fields if they would like, though we recommend calling the
+parent function to set up these fields as a base.
+
+:field text_vec:      bsz x seqlen tensor containing the parsed text data.
```

Comment:
I really appreciate this, thank you.

If we want to be really thorough, we could annotate with types too. (`:field Tensor[int64] text_vec`). This isn't really necessary.

Shame it isn't 2020 so that we have [Data Classes](https://docs.python.org/3/library/dataclasses.html#module-dataclasses).

Reply:
wow! I'm already starting to be ready to add types soon but we aren't quite yet forcing python 3.6 over some earlier 3.Xs